### ISSUE TITLE
Initial integration of ZenDNN as backend into PyTorch

### ISSUE THREAD

**Issue by ratan-prasad** [pytorch/pytorch#76244]
### 🚀 The feature, motivation and pitch

ZenDNN library optimizes basic meth kernels used by Deep learning algorithms on AMD EPYC CPUs. As part of integration of ZenDNN into PyTorch we integrated a single op (convolution). We have support for other ops and optimization ready to be submitted.

This PR covers following
- Added ZenDNN as third-party library
    - ZenDNN is a submodule for pytorch
    - modified pytorch build infra to build with ZenDNN when USE_ZENDNN is set
      to 1
    - pytorch build reverts to default behavior if USE_ZENDNN environmental
      variable is set to zero or unset
- Added blis (dependency of ZenDNN) as third-party library
    - blis is a submodule for pytorch
    - modified pytorch build infra to build with blis when BLAS is set to BLIS
    - blis is built if USE_ZENDNN is set to 1
- Added ZenDNN as backend to pytorch
    - added support for user level control to disable or enable zendnn backend
- Added following changes
    - added zendnn layout and ZendnnCPU dispatch key.
    - tensor conversion between dense and zendnn layouts
    - zendnn convolution as a new ConvBackend
        - when built with zendnn (USE_ZENDNN=1) and gradient calculations are
          disabled (torch.no_grad) convolution is computed using ZenDNN library)
        - backward propagation is not supported.
        - support is only added for fp32 data type
- Added few unit tests for pytorch + zendnn

URLs of dependent Library:
ZenDNN github: https://github.com/amd/ZenDNN
AMD-BLIS github: https://github.com/amd/blis
Corresponding Pull Request: https://github.com/pytorch/pytorch/pull/76242

**Comment by gchanan** [comment]
@emad can we assign someone to be the POC here?

**Comment by seemethere** [comment]
Do we have any CI machines that we can use to actively test this as well?

**Comment by b0noI** [comment]
AWS does have support of AMD EPYC CPUs. Random example EC2 M6a

**Comment by dagitses** [comment]
ping @emad. Can you assign a PoC here please.

**Comment by malfet** [comment]
What are the performance gain of using ZenDNN/BLIS over oneDNN/OpenBLAS on AMD EPYC CPU? And how about say SkyLake CPU? 

From PyTorch developer point of view, it would be ideal if framework yields best performance regardless of CPU architecture, which proposed feature does not seem to achieve, does it?

**Comment by ratan-prasad** [comment]
This PR contains only one op integrated to validate the procedure ahead of submitting the remaining PRs integrating the rest of the ZenDNN ops. To showcase the potential uplift using ZenDNN vs. Intel MKL-DNN on AMD CPU, we have benchmarked our full integrated version on PT v1.11. 
PyTorch v1.11 Throughput(images/sec) BS:640, Milan 64C NPS4

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 15">
<meta name=Originator content="Microsoft Word 15">
<link rel=File-List
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<link rel=themeData
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_themedata.thmx">
<link rel=colorSchemeMapping
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_colorschememapping.xml">





</head>

<body lang=EN-US style='tab-interval:.5in;word-wrap:break-word'>




Models | Improvement
-- | --
ResNet50 | 6.49%
GoogleNet | 118.83%
vgg11 | 18.17%
Inception-v4 | 69.96%
Alexnet | 17.99%




</body>

</html>

<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:w="urn:schemas-microsoft-com:office:word"
xmlns:m="http://schemas.microsoft.com/office/2004/12/omml"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Word.Document>
<meta name=Generator content="Microsoft Word 15">
<meta name=Originator content="Microsoft Word 15">
<link rel=File-List
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<link rel=themeData
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_themedata.thmx">
<link rel=colorSchemeMapping
href="file:///C:/Users/ratprasa/AppData/Local/Temp/msohtmlclip1/01/clip_colorschememapping.xml">





</head>

<body lang=EN-US style='tab-interval:.5in;word-wrap:break-word'>


<p class=MsoListParagraph style='margin-left:1.0in'>PyTorch v1.11 Latency(ms)
BS:1 Milan 64C NPS4<o:p></o:p></p>



Models | Improvement
-- | --
ResNet50 | 6.49%
GoogleNet | 118.83%
vgg11 | 18.17%
Inception-v4 | 69.96%
Alexnet | 17.99%




</body>

</html>

As ZenDNN library was designed to provide optimizations for AMD CPUs, it is not recommended to use ZenDNN with Intel CPUs. While models may still run, Intel CPUs will not achieve optimal performance with ZenDNN library, given that the optimizations were not tuned for their architecture. While we understand that it would be ideal if a framework yielded the best performance across all CPU architectures, we are proposing ZenDNN’s inclusion as an AMD CPU backend because we have not found currently available, comparable libraries to yield best performance for AMD CPU architecture.
